### PR TITLE
feat: custom tag

### LINKED ISSUE #3566: Feature: Tags to be propagated to container registry and auto-increment
### 🔖 Feature description
There are 2 propositions under this functionality:
1. TAGs that are created should be auto-incremented within a specified template. For example, the template v1.2.X is provided, the value of X should be auto-incremented and maintained by Devtron to generated tags like v1.2.0, v1.2.1, v1.2.2 and so on.
2. These TAGs should also be propagated to the container-images before pushing to the container registry.

### 🎤 Pitch

Maintaining TAGS and auto-incrementing it will help and ease Software delivery and Rollbacks. Propagating tags at Container registry will help enable effective clean up policies at Container Repository level and save Cloud costs. 

### ✌️ Solution

-

### 🔄️ Alternative

NA

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/devtron-labs/devtron/blob/main/CODE_OF_CONDUCT.md)

AB#3988

### PR DESCRIPTION
**Pull request by avdhesh-devtron**
feat: custom tag
Fixes #3566 

## How Has This Been Tested?
- dev testing
- QA testing

## Checklist:

* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?
Yes
```release-note

```

### CODE REVIEW

File: api/bean/CustomTag.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package bean
+
+type CustomTag struct {
+	EntityKey            int    `json:"entityKey"`
+	EntityValue          string `json:"entityValue"`
+	TagPattern           string `json:"tagPattern"`
+	AutoIncreasingNumber int    `json:"counterX"`
```
Comment (vikramdevtron) [inline api/bean/CustomTag.go:7]: can be used better name for counterX

File: pkg/CustomTagService.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+	"strings"
+)
+
+const (
+	EntityNull = iota
+	EntityTypeCiPipelineId
+)
+
+const (
+	imagePathPattern                                              = "%s/%s:%s" // dockerReg/dockerRepo:Tag
```
Comment (vikramdevtron) [inline pkg/CustomTagService.go:20]: we can put these type of constants at some common place

File: pkg/CustomTagService.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+func (impl *CustomTagServiceImpl) CreateOrUpdateCustomTag(tag *bean.CustomTag) error {
+	if err := validateTagPattern(tag.TagPattern); err != nil {
+		return err
+	}
+	customTagData := repository.CustomTag{
+		EntityKey:            tag.EntityKey,
+		EntityValue:          tag.EntityValue,
+		TagPattern:           strings.ReplaceAll(tag.TagPattern, "{X}", "{x}"),
```
Comment (vikramdevtron) [inline pkg/CustomTagService.go:63]: can we use some other relevant name than X

File: pkg/CustomTagService.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+func validateTagPattern(customTagPattern string) error {
+	if len(customTagPattern) == 0 {
+		return fmt.Errorf("tag length can not be zero")
+	}
+
+	// for patterns like v1.0.{x} we will calculate count with . in {x} i.e .{x}
+	variableCount := 0
+	variableCount = variableCount + strings.Count(customTagPattern, "{x}")
```
Comment (vikramdevtron) [inline pkg/CustomTagService.go:153]: can we use some other placeholder everywhere for x

File: pkg/pipeline/CiHandler.go [hunk 5]
```diff
@@ -1663,9 +1694,12 @@ func (impl *CiHandlerImpl) UpdateCiWorkflowStatusFailure(timeoutForFailureCiBuil
 			err := impl.ciWorkflowRepository.UpdateWorkFlow(ciWorkflow)
 			if err != nil {
 				impl.Logger.Errorw("unable to update ci workflow, its eligible to mark failed", "err", err)
-				continue
 				// skip this and process for next ci workflow
 			}
+			err = impl.customTagService.DeactivateImagePathReservation(ciWorkflow.ImagePathReservationId)
+			if err != nil {
+				impl.Logger.Errorw("unable to update ci workflow, its eligible to mark failed", "err", err)
```
Comment (vikramdevtron) [inline pkg/pipeline/CiHandler.go:1701]: why not returning error

File: pkg/pipeline/WebhookService.go [hunk 6]
```diff
@@ -141,6 +145,13 @@ func (impl WebhookServiceImpl) HandleCiStepFailedEvent(ciPipelineId int, request
 		return err
 	}
 
+	go func() {
```
Comment (vikramdevtron) [inline pkg/pipeline/WebhookService.go:148]: can use use nats for this purpose

File: scripts/sql/182_custom_image_tag.down.sql [hunk 7]
```diff
@@ -0,0 +1,15 @@
+DROP TABLE IF EXISTS custom_tag;
```
Comment (vikramdevtron) [inline scripts/sql/182_custom_image_tag.down.sql:1]: check seq number with latest main

File: pkg/pipeline/CiHandler.go (hunk 5 above)
Comment (iamayushm) [reply pkg/pipeline/CiHandler.go:1701]: there are more than one ci_workflows updated in this function. If there is some error with one workflow then we are skipping only that workflow and continuing our loop

**Review by vikramdevtron — APPROVED** [review]

**Pull request merged**